modules/reindex/src/main/java/org/elasticsearch/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -819,147 +815,99 @@ public static RequestWrapper<DeleteRequest> wrap(DeleteRequest request) {
+        protected Map<String, Object> extractSource(Map<String, Object> sourceWrapper) {
+            Object rawSource = sourceWrapper.get("_source");
+            if (rawSource == null) {
+                throw new IllegalArgumentException("missing source");
+            } else if (sourceWrapper.size() > 1) {
+                throw new IllegalArgumentException(
+                    "Invalid fields added to context ["
+                        + sourceWrapper.keySet()
+                            .stream()
+                            .filter(k -> "_source".equals(k) == false)
+                            .sorted()
+                            .collect(Collectors.joining(", "))
+                        + "]"
+                );
             }
-
-            if (false == context.isEmpty()) {
-                throw new IllegalArgumentException("Invalid fields added to context [" + String.join(",", context.keySet()) + ']');
+            if (rawSource instanceof Map<?, ?> source) {
+                return (Map<String, Object>) source;
             }
-            return request;
+            throw new IllegalArgumentException(
+                "unexpected type for _source [" + rawSource + "] expected Map, but got [" + rawSource.getClass().getName() + "]"
+            );
         }
 
-        protected RequestWrapper<?> scriptChangedOpType(RequestWrapper<?> request, OpType oldOpType, OpType newOpType) {
-            switch (newOpType) {
-                case NOOP -> {
+        protected RequestWrapper<?> requestFromOp(RequestWrapper<?> request, String op) {
+            switch (op) {
+                case "noop" -> {
                     taskWorker.countNoop();
                     return null;
                 }
-                case DELETE -> {
+                case "delete" -> {
                     RequestWrapper<DeleteRequest> delete = wrap(new DeleteRequest(request.getIndex(), request.getId()));
                     delete.setVersion(request.getVersion());
                     delete.setVersionType(VersionType.INTERNAL);
                     delete.setRouting(request.getRouting());
                     return delete;
                 }
-                default -> throw new IllegalArgumentException(
-                    "Unsupported operation type change from [" + oldOpType + "] to [" + newOpType + "]"
-                );
+                case INDEX -> {
+                    return request;
+                }
+                default -> throw new IllegalArgumentException("Unsupported operation type change from [" + INDEX + "] to [" + op + "]");
             }
         }
 
-        protected abstract void scriptChangedIndex(RequestWrapper<?> request, Object to);
-
-        protected abstract void scriptChangedId(RequestWrapper<?> request, Object to);
-
-        protected abstract void scriptChangedVersion(RequestWrapper<?> request, Object to);
-
-        protected abstract void scriptChangedRouting(RequestWrapper<?> request, Object to);
```

Comment:
Another question for my clarity -- these are removed because the onus is now on metadata to error at the time it's changed using the validators as opposed to after the fact? And then updateRequest will actually update the RequestWrapper upon script completion?

Reply:
Yup, that's correct.  These enforced read-only in Update, which is now the job of UpdateMetadata. The non-empty implementation of `updateRequest` in `Reindexer` handles these for Reindex.